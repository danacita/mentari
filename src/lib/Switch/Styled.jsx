import styled from 'styled-components';
import Color from '../Styles/bases/Color';

const SwitchWrapper = styled.label`
  position: relative;
  display: inline-block;
  width: 48px;
  height: 24px;
`;

const SwitchInput = styled.input`
  opacity: 0;
  width: 0;
  height: 0;
  &:checked {
    + span {
      background-color: ${Color.blue};
      &:before {
        -webkit-transform: translateX(26px);
        -ms-transform: translateX(26px);
        transform: translateX(26px);
        border: 2px solid ${Color.blue};
      }
    }
  }
`;

const SwitchSlider = styled.span`
  position: absolute;
  cursor: pointer;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: #ccc;
  -webkit-transition: 0.4s;
  transition: 0.4s;
  border-radius: 24px;
  &:before {
    position: absolute;
    border-radius: 50%;
    border: 2px solid ${Color.grey191};
    content: '';
    height: 24px;
    width: 24px;
    left: 0px;
    bottom: 0px;
    background-color: white;
    -webkit-transition: 0.4s;
    transition: 0.4s;
  }
`;

export { SwitchWrapper, SwitchInput, SwitchSlider };
