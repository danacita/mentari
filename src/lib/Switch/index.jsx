import PropTypes from 'prop-types';
import React from 'react';
import { SwitchWrapper, SwitchInput, SwitchSlider } from './Styled';

const propTypes = {
  disabled: PropTypes.bool,
  customOnChange: PropTypes.func,
  // input prop is passed by redux-form to keep field in sync with state
  input: PropTypes.shape({
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.bool]),
    onChange: PropTypes.func
  }).isRequired
};

const defaultProps = {
  disabled: false,
  customOnChange: () => {}
};

const SwitchField = ({ disabled, customOnChange, input }) => (
  <React.Fragment>
    <SwitchWrapper>
      <SwitchInput
        type="checkbox"
        disabled={disabled}
        onChange={value => {
          // Custom onChange function to make form dynamic
          customOnChange(value);
          // Redux Form onChange function to sync field value with state
          input.onChange(value);
        }}
      />
      <SwitchSlider />
    </SwitchWrapper>
  </React.Fragment>
);

SwitchField.propTypes = propTypes;
SwitchField.defaultProps = defaultProps;

export default SwitchField;
