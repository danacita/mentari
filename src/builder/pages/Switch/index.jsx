import React, { Component } from 'react';
import SyntaxHighlighter from 'react-syntax-highlighter';
import { atomOneDarkReasonable } from 'react-syntax-highlighter/dist/esm/styles/hljs';
import { ContentWithRightBar } from '../Styled';
import { Margin, SwitchField, Checkbox } from '../../../lib';
import PropsShared from '../../shared/Props';
import DemoShared from '../../shared/Demo';
import CodeShared from '../../shared/Code';
import RightbarShared from '../../shared/Rightbar';

class SwitchPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      handleSwitchDisable: false
    };
  }

  render() {
    const { handleSwitchDisable } = this.state;
    const codeString = `import { SwitchField } from '@danacita/mentari';\n<SwitchField ${
      handleSwitchDisable ? `disabled ` : ''
    }/>`;
    return (
      <ContentWithRightBar>
        <h1>Switch</h1>
        <p>Lorem ipsum dolor, sit amet consectetur adipisicing elit.</p>
        <Margin top={24}>
          <PropsShared props>
            <li>
              <span>disabled</span>
              <span>Apply disabled state of Switch for unusable and unclickable</span>
            </li>
          </PropsShared>
        </Margin>
        <DemoShared>
          <h4>Demo</h4>
          <p>You can change the props via the component builder on your right</p>
          <Margin top={16}>
            <SwitchField withRedux={handleSwitchDisable} disabled={handleSwitchDisable} />
          </Margin>
        </DemoShared>
        <h4>Code</h4>
        <Margin top={16}>
          <CodeShared copyCode={codeString}>
            <SyntaxHighlighter language="jsx" style={atomOneDarkReasonable} showLineNumbers>
              {codeString}
            </SyntaxHighlighter>
          </CodeShared>
        </Margin>
        <RightbarShared title="Textfield">
          <Margin bottom={16}>
            <h4>State of Switch</h4>
          </Margin>
          <Margin bottom={16}>
            <Checkbox
              id="handleSwitchDisable"
              name="handleSwitchDisable"
              label="Disable"
              checked={handleSwitchDisable}
              withRedux={false}
              onChange={e => this.setState({ handleSwitchDisable: e.target.checked })}
            />
          </Margin>
        </RightbarShared>
      </ContentWithRightBar>
    );
  }
}

export default SwitchPage;
