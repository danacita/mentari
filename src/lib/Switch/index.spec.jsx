import React from 'react';
import renderer from 'react-test-renderer';

import Switch from '.';

describe('Switch', () => {
  it('renders correctly', () => {
    const input = {
      value: 'Value',
      onChange: () => {}
    };

    const meta = {
      error: '',
      touched: false
    };

    const tree = renderer.create(<Switch input={input} meta={meta} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
