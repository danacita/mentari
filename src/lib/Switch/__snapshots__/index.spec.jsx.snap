// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch renders correctly 1`] = `
<label
  className="sc-bdVaJa jBqTSZ"
>
  <input
    className="sc-bwzfXH gbvsbZ"
    disabled={false}
    onChange={[Function]}
    type="checkbox"
  />
  <span
    className="sc-htpNat kLbBuw"
  />
</label>
`;
